src/Symfony/Component/Validator/Mapping/ConstraintChecker.php

```diff
@@ -0,0 +1,46 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Validator\Mapping;
+
+/**
+ * Helper methods to check if a property or a class has a given constraint.
+ *
+ * @internal
+ *
+ * @author Kévin Dunglas <dunglas@gmail.com>
+ */
+trait ConstraintChecker
+{
+    private function propertyHasConstraint(ClassMetadata $metadata, string $constraintType, string $fieldName): bool
+    {
+        foreach ($metadata->getPropertyMetadata($fieldName) as $propertyMetadata) {
+            foreach ($propertyMetadata->getConstraints() as $constraint) {
+                if (is_a($constraint, $constraintType, true)) {
+                    return true;
+                }
+            }
+        }
+
+        return false;
+    }
+
+    private function classHasConstraint(ClassMetadata $metadata, string $constraintType): bool
+    {
+        foreach ($metadata->getConstraints() as $constraint) {
```

Comment:
You have the exact same loop on the method above, you can simplify :)

Reply:
I can introduce a new private method, but I'll add a new function call in O(n2) loops... I don't think it is worth it.